### PR TITLE
Add SRv6 SID L3Adj HLD

### PR DESCRIPTION
**Pull request by cscarpitta**
This HLD describes the extensions of SRv6Orch required to support the programming of the L3Adj associated with SRv6 uA, End.X, uDX4, uDX6, End.DX4, and End.DX6 behaviors.

The document that describes the HLD is available here:
https://github.com/cscarpitta/SONiC/blob/srv6_sid_l3adj/doc/srv6/srv6_sid_l3adj.md

| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-swss | [[orchagent]: Extend the SRv6Orch to support the programming of the L3Adj](https://github.com/sonic-net/sonic-swss/pull/2902) | GitHub issue/pull request detail |

### CODE REVIEW

File: doc/srv6/srv6_sid_l3adj.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+- SRv6Orch invokes the sairedis `sai_srv6_api->create_my_sid_entry()` API to create the SRv6 SID entry in the ASIC DB. 
+
+The next subsections describe the SRv6Orch changes required to support the HLD described above.
+
+### SRv6Orch Changes
+
+We extend SRv6Orch to support the programming of the L3Adj associated with uA, End.X, uDX4, uDX6, End.DX4, and End.DX6 behaviors.
+
+When SRv6Orch receives a SID with the `adj` parameter set, it calls the function `neighOrch->hasNextHop()` to make sure a nexthop associated with the adjacency exists.
+- If the nexthop does not exist, SRv6Orch returns an error.
```
Comment (kperumalbfn) [inline doc/srv6/srv6_sid_l3adj.md:74]: How about ECMP? Does Fpmsyncd change(https://github.com/sonic-net/sonic-swss/pull/2515/files) handle ECMP as well? Please update accordingly.

**Comment by zhangyanzhao** [comment]
community review recording https://zoom.us/rec/share/KXIvrUaqMzYGe1rXnIwNkUrxQ6LM9rE9vqP46Xo1eQYT67kNWDEn3b9o4JyBMXWk.nkzFHaTnToC2mr7U

**Comment by shuaishang** [comment]
@cscarpitta 
SRv6Orch need to handle the scenarios that nexthop is not exist or re-created:
1. when configure L3Adj, the nexthop is not exist. (e.g. when system reboot, L3Adj is configured but nexthop is not exist yet)
2. nexthop is removed/added because of neighbour down/up

You can refer to "mirrororch.cpp" which is consumer of nexthop.

**Comment by cscarpitta** [comment]
@shuaishang Thanks for the comment. We added a callback to handle neighbor "up" / "down" events, as discussed during the community review.

**Comment by ahsalam** [comment]
This PR has been reviewed in Routing WG weekly meeting - Thursday, September 28, 2023. 

Here is the summary of agreement reached during the call 

image

https://lists.sonicfoundation.dev/g/sonic-wg-routing/message/106

We updated the HLD and the code PR as per the above  agreement. 

@venkatmahalingam, since you there in the Routing WG weekly meeting. Could you please review and help to merge the HLD and the code PR. Thanks

**Comment by ahsalam** [comment]
@kperumalbfn  could you please help and review this PR?
This feature is tracked for 202311 release.
@zhangyanzhao

File: doc/srv6/srv6_sid_l3adj.md (hunk 1 above)
Comment (kperumalbfn) [reply doc/srv6/srv6_sid_l3adj.md]: @cscarpitta Could you update the doc with ECMP for MYSID entries.

Comment (ahsalam) [reply doc/srv6/srv6_sid_l3adj.md]: @ kperumalbfn Hi Kumaresh, the scope of the PR will be limited to a single L3Adj. We are doing of the required changes to support ECMP and we are targeting 202405 for the ECMP support. Thanks!

**Comment by eddieruan-alibaba** [comment]
This proposal has been reviewed in routing WG. 

https://lists.sonicfoundation.dev/g/sonic-wg-routing/wiki/34965

Current version looks good to me.

**Review by eddieruan-alibaba — APPROVED** [review]
It has been reviewed in Routing WG.

**Comment by ahsalam** [comment]
Hi @zhangyanzhao @kperumalbfn 

All the comments of this PR has been addressed. It has been reviewed in Routing WG. Eddie has approved the PR. Could we merge the PR ?

**Review by kperumalbfn — APPROVED** [review]

**Pull request merged**